### ISSUE TITLE
Why do we need to specify key in `botocore.signers.generate_presigned_post`?

### ISSUE THREAD

**Issue by ahubers** [boto/botocore#1481]
## The Issue

The function signature found here https://github.com/boto/botocore/blob/master/botocore/signers.py#L600

reads
```python
def generate_presigned_post(self, Bucket, Key, Fields=None, Conditions=None,
                            ExpiresIn=3600):
```

which would imply that the `Key` is mandatory when generating a presigned POST policy. However the [AWS reference](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html) doesn't seem to have that constraint. 

## My Motivation

I'm asking this here because I'd like to generate a POST policy for mobile clients that will allow them to perform multiple uploads without the need to request a new policy per upload. My thought being that I can do this fairly easily by only enforcing a "starts-with"  condition on the S3 key. So my policy would look more like the example policy AWS gives in the above reference doc:

```json
{ "expiration": "2007-12-01T12:00:00.000Z",
  "conditions": [
    {"acl": "public-read" },
    {"bucket": "johnsmith" },
    ["starts-with", "$key", "user/eric/"],
  ]
}
```

E.g, I could use this set of conditionals to generate a POST policy which a mobile client could use to upload a whole album of photos.

## TL;DR

When including `Key` as a required argument in the function signature of `generate_presigned_post`, it implies I need to generate a new presigned S3 post for every additional client upload which is bothersome / impossible if I want to do batch uploading.

So I am requesting that become a kwarg. 

Let me know if I'm making sense. Especially let me know if I'm just an idiot who missed something. That happens a lot with me.

**Comment by ahubers** [comment]
>The good news is that it should be possible to relax the constraint in boto3.
I'm a bit unfamiliar with the boto structure. Are you saying it should be possible to relax this constraint as a future feature of boto3, or by using boto3 as opposed to botocore, I can avoid this constraint?

I feel it's probably the former... From what I understand Boto3 just offers an interface to expose botocore service clients?

**Comment by ahubers** [comment]
If there is a correct place for me to modify this and open a PR let me know... Obviously I can switch def `generate_presigned_post(self, Bucket, Key, ` to `def generate_presigned_post(self, Bucket, Key=None` and work from there, but I'm not sure how that fares with respect to backwards compatibility and/or annoying devs using boto3.

Thanks!

**Comment by birdbrained** [comment]
Looking at the code here, I don't see the `generate_presigned_url` method having a `Key` kwarg. Is that an older version of botocore that had that?
> We made key required to parallel the interface with the presign url functionality and may have not been aware of the ability to use startswith instead of key.

**Comment by madhurjain** [comment]
Can be achieved by -

```py
s3_client.generate_presigned_post("bucketname", "${filename}")
```
Or
```py
s3_client.generate_presigned_post("bucketname", "user/eric/${filename}")
```
Because looking at the code for `botocore/signers.py`, ending the key argument with `${filename}` makes it optional and instead adds a `starts-with` condition
```
# If the key ends with filename, the only constraint that can be
# imposed is if it starts with the specified prefix.
if key.endswith('${filename}'):
    conditions.append(["starts-with", '$key', key[:-len('${filename}')]])
else:
    conditions.append({'key': key})
```
